src/Storefront/Framework/SystemCheck/ProductDetailReadinessCheck.php

```diff
@@ -0,0 +1,153 @@
+use Shopware\Core\Framework\Uuid\Uuid;
+use Shopware\Storefront\Framework\SystemCheck\Util\SalesChannelDomainUtil;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @internal
+ */
+#[Package('discovery')]
+class ProductDetailReadinessCheck extends BaseCheck
+{
+    private const DETAIL_PAGE = 'frontend.detail.page';
+
+    private const MESSAGE_SUCCESS = 'Product detail pages are OK for provided sales channels.';
+
+    private const MESSAGE_FAILURE = 'Some or all product detail pages are unhealthy.';
+
+    public function __construct(
+        private readonly SalesChannelDomainUtil $util,
+        private readonly Connection $connection,
+    ) {
+    }
+
+    public function run(): Result
+    {
+        return $this->util->runAsSalesChannelRequest(
+            fn () => $this->util->runWhileTrustingAllHosts(
+                fn () => $this->doRun()
+            )
+        );
+    }
+
+    public function category(): Category
+    {
+        return Category::FEATURE;
+    }
+
+    public function name(): string
+    {
+        return 'ProductDetailReadiness';
+    }
+
+    protected function allowedSystemCheckExecutionContexts(): array
+    {
+        return SystemCheckExecutionContext::readiness();
+    }
+
+    private function doRun(): Result
+    {
+        $domains = $this->fetchSalesChannelDomains();
+        $salesChannelIds = array_keys($domains);
+        $productIds = $salesChannelIds ? $this->fetchProductIds($salesChannelIds) : null;
+
+        $extra = [];
+        $requestStatus = [];
+        foreach ($domains as $salesChannelId => $domain) {
+            $productId = $productIds[$salesChannelId] ?? null;
+
+            if ($productId === null) {
+                continue;
+            }
+
+            $url = $this->util->generateDomainUrl($domain, self::DETAIL_PAGE, [
+                'productId' => $productId,
+            ]);
+
+            $request = Request::create($url);
+            $result = $this->util->handleRequest($request);
+
+            $status = $result->responseCode >= Response::HTTP_BAD_REQUEST ? Status::FAILURE : Status::OK;
+            $requestStatus[$status->name] = $status;
+
+            $extra[] = $result->getVars();
+        }
+
+        if ($requestStatus === []) {
+            return $this->util->createEmptyResult($this->name(), 'No sales channels with product detail pages found.');
+        }
+
+        $finalStatus = \count($requestStatus) === 1 ? current($requestStatus) : Status::ERROR;
+
+        return new Result(
+            $this->name(),
+            $finalStatus,
+            $finalStatus === Status::OK ? self::MESSAGE_SUCCESS : self::MESSAGE_FAILURE,
+            $finalStatus === Status::OK,
+            $extra
+        );
+    }
+
+    /**
+     * @param list<string> $salesChannelIds
+     *
+     * @return array<string, string>
+     */
+    private function fetchProductIds(array $salesChannelIds): array
```

Comment:
So we're fetching and testing ALL products? I mean it has the nice benefit of pre warming the cache, but is it not a little excessive with stores that have more than 10k products? Doing this serially will also be pretty slow

Reply:
Using the seo urls would also be a lot simpler.

Reply:
@pweyck we only fetch the key pair, so we test one product per sales channel ;)

Reply:
Makes sense 🙈 Need more ☕